app/src/main/java/com/google/android/sambadocumentsprovider/ShareManager.java

```diff
@@ -56,19 +63,41 @@
   private final Set<String> mMountedServerSet = new HashSet<>();
   private final Map<String, String> mServerStringMap = new HashMap<>();
   private final CredentialCache mCredentialCache;
+  private final EncryptionManager mEncryptionManager;
 
   private final List<MountedShareChangeListener> mListeners = new ArrayList<>();
 
   ShareManager(Context context, CredentialCache credentialCache) {
     mCredentialCache = credentialCache;
 
+    mEncryptionManager = SambaProviderApplication.getEncryptionManager(context);
+
     mPref = context.getSharedPreferences(SERVER_CACHE_PREF_KEY, Context.MODE_PRIVATE);
     // Loading saved servers.
-    final Set<String> serverStringSet =
+    Set<String> serverStringSet =
         mPref.getStringSet(SERVER_STRING_SET_KEY, Collections.<String> emptySet());
+
+    if (!mEncryptionManager.isKeyRegistered(ENCRYPTION_KEY_ALIAS)) {
```

Comment:
Don't expose key outside EncryptionManager. Save the key alias inside of it. `decrypt()` can throw the exception if it just fails, and `ShareManager` can then try to load the content as plain text.

If you still want to keep its capability to generate multiple keys, change the name of `ENCRYPTION_KEY_ALIAS` to `ENCRYPTION_TOKEN` and pass it in `encrypt()` and `decrypt()`. Generate key on demand in `encrypt()` method.